api/src/main/java/com/example/api/service/user/PasswordResetService.java

```diff
@@ -0,0 +1,81 @@
+import com.example.api.service.validator.UserValidator;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import javax.transaction.Transactional;
+import java.util.UUID;
+
+@Service
+@RequiredArgsConstructor
+@Slf4j
+@Transactional
+public class PasswordResetService {
+    private final int TOKEN_SIZE = 8;
+    private final int TOKEN_TTL_IN_MIN = 5;
+
+    private final UserValidator userValidator;
+    private final UserRepo userRepo;
+    private final EmailService emailService;
+    private final PasswordEncoder passwordEncoder;
+    private final PasswordResetValidator passwordResetValidator;
+    private final PasswordResetTokenRepo passwordResetTokenRepo;
+    private final PasswordValidator passwordValidator;
+
+    public void sendPasswordResetEmail(String email) {
+        User user = userRepo.findUserByEmail(email);
+        userValidator.validateUserIsNotNull(user, email);
+
+        String plainToken = UUID.randomUUID()
+                .toString()
+                .replaceAll("_", "")
+                .substring(0, TOKEN_SIZE);
+        String hashedToken = passwordEncoder.encode(plainToken);
+
+        PasswordResetToken token = new PasswordResetToken();
+        token.setIsUsed(false);
+        token.setHashedToken(hashedToken);
+        token.setCreationTime(System.currentTimeMillis());
+        token.setExpirationTime(token.getCreationTime() + (TOKEN_TTL_IN_MIN * 60 * 1_000));
+        token.setUser(user);
+        passwordResetTokenRepo.save(token);
+        user.setPasswordResetToken(token);
+        userRepo.save(user);
+
+        log.info("Sending password reset email to {}", email);
+        emailService.sendEmail(user, "Systematic chaos - password reset", getPasswordResetEmailMessage(plainToken));
+
+    }
+
+    public void resetPassword(ResetPasswordForm form) throws RequestValidationException {
+        log.info("Trying to reset password for {}", form.getEmail());
+        String email = form.getEmail();
+        User user = userRepo.findUserByEmail(email);
+        userValidator.validateUserIsNotNull(user, email);
+
+        PasswordResetToken token = user.getPasswordResetToken();
+        passwordResetValidator.validateTokenIsNotNull(email, token);
+        passwordResetValidator.validateTokenIsNotUsed(email, token);
+        passwordResetValidator.validateExpirationTime(email, System.currentTimeMillis(), token.getExpirationTime());
+        passwordResetValidator.validateToken(email, passwordEncoder, form.getPasswordResetToken(), token.getHashedToken());
+        passwordValidator.validatePassword(form.getNewPassword());
+        user.setPassword(passwordEncoder.encode(form.getNewPassword()));
+        token.setIsUsed(true);
+    }
+
+    private String getPasswordResetEmailMessage(String token) {
+        return String.format("Your password reset token: %s", token);
```

Comment:
Czy ta wiadomość musi mieć max długość? Jeżeli nie (lub jeżeli tak ale to poniżej jest wystarczająco krótkie) to zamień na coś takiego (może być po polsku wsm xD):
```
Hello,

Systematic Chaos website administration here. You are receiving this email because it was provided for password reset on our website.

Your password change token is: **TOKEN HERE**

If you have not provided your e-mail address or do not know why you received this message - ignore it.

The message was generated automatically, do not reply to it.

Best regards,
Systematic Chaos Administration
```

Reply:
Zrobione